csrc/device_lower/analysis/tensor_memory.cpp

```diff
@@ -214,6 +250,133 @@ computeTMemLdStDataPath(Fusion* fusion) {
     } else {
       continue;
     }
+    const auto& loop_domain = ir_utils::getTvOutput(ldst)->getLoopDomain();
+    const auto& tmem_tv_info = allocation.getTVInfo(tmem_tv);
+    auto& id_graph = GpuLower::current()->tensorIndexer().traversalGraph();
+    ValGroups lane_allocation_valgroups =
+        id_graph.toGroups(tmem_tv_info.lane_allocation);
+
+    NVF_CHECK(
+        !tid_ptypes.empty(),
+        "Invalid data access pattern in TMem load/store: ",
+        "TMem load/store must be warp-collective, but CTA size is one.");
+
+    // We need to construct a ValGroup that represents "warp" for this
+    // expression from consumer's loop domain. Naively speaking, it is just
+    // split(TIDz * TIDy * TIDx, 32).inner, where TIDz, TIDy and TIDx are the
+    // IterDomains in the loop domain that has such parallelization. But
+    // unfortunately, in reality, it is not that simple. NVFuser allows
+    // parallelizating IterDomains in an inexact way, for example, if the
+    // kernel's parallel dimension size for TIDx is 64, then the IterDomain
+    // being parallelized with TIDx does not have to be exactly 64. This
+    // inexactness is especially common in warp-specialized kernels. If, for
+    // example, the TIDx parallelized IterDomain in the loop domain is not
+    // exact, then split(TIDz * TIDy * TIDx, 32).inner may not be the warp.
+    // To handle this, we need to introduce a concept called "contiguity of
+    // parallel types in the loop domain". We can represent wap as
+    // split(TIDz * TIDy * TIDx, 32).inner if and only if TIDz and TIDy
+    // are contiguous. If TIDz is not contiguous but TIDy is, then warp would
+    // be split(TIDy * TIDx, 32).inner. If neither TIDz nor TIDy is contiguous,
+    // then warp would be split(TIDx, 32).inner.
+
+    // Get the contiguity of tid_ptypes in the loop domain.
+    // The contiguity of each item in tid_ptypes are defined as follows:
+    // - The inner tid_ptypes is always contiguous.
+    // - The item at index i is contiguous if the item at index i+1 is
+    //   exact(its extent in the loop domain is the same as parallel
+    //   dimension size of the kernel).
```

Comment:
Even just a few examples would be appreciated.

Reply:
Added